packages/iov-core/src/jsonrpcsigningserver.ts

```diff
@@ -0,0 +1,185 @@
+}
+
+function parseRpcCall(data: JsonRpcRequest): RpcCall {
+  if (!isJsonCompatibleDictionary(data.params)) {
+    throw new Error("Request params are only supported as dictionary");
+  }
+
+  switch (data.method) {
+    case "getIdentities": {
+      const { reason, chainIds } = data.params;
+      if (typeof reason !== "string") {
+        throw new ParamsError("1st parameter (reason) must be a string");
+      }
+      if (!Array.isArray(chainIds)) {
+        throw new ParamsError("2nd parameter (chainIds) must be an array");
+      }
+      if (!isArrayOfStrings(chainIds)) {
+        throw new ParamsError("Found non-string element in chainIds array");
+      }
+      return {
+        name: "getIdentities",
+        reason: reason,
+        chainIds: chainIds,
+      };
+    }
+    case "signAndPost": {
+      const { reason, transaction } = data.params;
+      if (typeof reason !== "string") {
+        throw new ParamsError("1st parameter (reason) must be a string");
+      }
+      if (typeof transaction !== "object") {
+        throw new ParamsError("2nd parameter (transaction) must be an object");
+      }
+      const parsedTransaction = TransactionEncoder.fromJson(transaction);
+      if (!isUnsignedTransaction(parsedTransaction)) {
+        throw new ParamsError("2nd parameter (transaction) does not look like an unsigned transaction");
+      }
+      return {
+        name: "signAndPost",
+        reason: reason,
+        transaction: parsedTransaction,
+      };
+    }
+    default:
+      throw new MethodNotFoundError("Unknown method name");
+  }
+}
+
+/**
+ * A transport-agnostic JSON-RPC wrapper around SigningServerCore
+ */
+export class JsonRpcSigningServer {
+  private readonly core: SigningServerCore;
+
+  constructor(core: SigningServerCore) {
+    this.core = core;
+  }
+
+  public async handleUnchecked(request: unknown): Promise<JsonRpcResponse> {
+    let checkedRequest: JsonRpcRequest;
+    try {
+      checkedRequest = parseJsonRpcRequest(request);
+    } catch (error) {
+      const requestId = parseJsonRpcId(request);
+      const errorResponse: JsonRpcErrorResponse = {
+        jsonrpc: "2.0",
+        id: requestId,
+        error: {
+          code: jsonRpcCode.invalidRequest,
+          message: error.toString(),
+        },
+      };
+      return errorResponse;
+    }
+
+    return this.handleChecked(checkedRequest);
+  }
+
+  /**
+   * Handles a checked JsonRpcRequest
+   *
+   * 1. convert JsRpcRequest into calls to SigningServerCore
+   * 2. call SigningServerCore
+   * 3. convert result to JSON-RPC format
+   */
+  public async handleChecked(request: JsonRpcRequest): Promise<JsonRpcResponse> {
+    let call: RpcCall;
+    try {
+      call = parseRpcCall(request);
+    } catch (error) {
+      const errorResponse: JsonRpcErrorResponse = {
+        jsonrpc: "2.0",
+        id: request.id,
+        error: {
+          code: jsonRpcCode.methodNotFound,
```

Comment:
What if the error is a `ParamsError`?

Reply:
Then we return jsonRpcCode.methodNotFound – why not! Nice catch